math/vecops/inc/ROOT/RVec.hxx

```diff
@@ -94,32 +111,985 @@ auto MapFromTuple(Tuple_t &&t, std::index_sequence<Is...>)
+   {
+      // Just do a copy.
+      uninitialized_copy(I, E, Dest);
+   }
+
+   /// Copy the range [I, E) onto the uninitialized memory
+   /// starting with "Dest", constructing elements into it as needed.
+   template <typename It1, typename It2>
+   static void uninitialized_copy(It1 I, It1 E, It2 Dest)
+   {
+      // Arbitrary iterator types; just use the basic implementation.
+      std::uninitialized_copy(I, E, Dest);
+   }
+
+   /// Copy the range [I, E) onto the uninitialized memory
+   /// starting with "Dest", constructing elements into it as needed.
+   template <typename T1, typename T2>
+   static void uninitialized_copy(
+      T1 *I, T1 *E, T2 *Dest,
+      typename std::enable_if<std::is_same<typename std::remove_const<T1>::type, T2>::value>::type * = nullptr)
+   {
+      // Use memcpy for PODs iterated by pointers (which includes SmallVector
+      // iterators): std::uninitialized_copy optimizes to memmove, but we can
+      // use memcpy here. Note that I and E are iterators and thus might be
+      // invalid for memcpy if they are equal.
+      if (I != E)
+         memcpy(reinterpret_cast<void *>(Dest), I, (E - I) * sizeof(T));
+   }
+
+   /// Double the size of the allocated memory, guaranteeing space for at
+   /// least one more element or MinSize if specified.
+   void grow(size_t MinSize = 0)
+   {
+      this->grow_pod(MinSize, sizeof(T));
+   }
+
+public:
+   using iterator = typename SuperClass::iterator;
+   using const_iterator = typename SuperClass::const_iterator;
+   using reference = typename SuperClass::reference;
+   using size_type = typename SuperClass::size_type;
+
+   void push_back(const T &Elt)
+   {
+      if (R__unlikely(this->size() >= this->capacity()))
+         this->grow();
+      memcpy(reinterpret_cast<void *>(this->end()), &Elt, sizeof(T));
+      this->set_size(this->size() + 1);
+   }
+
+   void pop_back() { this->set_size(this->size() - 1); }
+};
+
+/// Storage for the SmallVector elements.  This is specialized for the N=0 case
+/// to avoid allocating unnecessary storage.
+template <typename T, unsigned N>
+struct SmallVectorStorage {
+   alignas(T) char InlineElts[N * sizeof(T)]{};
+};
+
+/// We need the storage to be properly aligned even for small-size of 0 so that
+/// the pointer math in \a SmallVectorTemplateCommon::getFirstEl() is
+/// well-defined.
+template <typename T>
+struct alignas(T) SmallVectorStorage<T, 0> {
+};
+
+/// The size of the inline storage of an RVec.
+/// Our policy is to allocate at least 8 elements (or more if they all fit into one cacheline)
+/// unless the size of the buffer with 8 elements would be over a certain maximum size.
+template <typename T>
+struct RVecInlineStorageSize {
+private:
+#ifdef R__HAS_HARDWARE_INTERFERENCE_SIZE
+   constexpr std::size_t cacheLineSize = std::hardware_destructive_interference_size;
+#else
+   // safe bet: assume the typical 64 bytes
+   static constexpr std::size_t cacheLineSize = 64;
```

Comment:
Is it really safe? The A64FX has 256bytes cachlines: https://www.fujitsu.com/downloads/SUPER/a64fx/a64fx_datasheet.pdf
But I guess you know your clients better and I do :)

Reply:
as safe as it gets, but we will eventually always use `std::hardware_destructive_interference_size`, users "just" need to get compilers that implement it

Reply:
Using cmake configuration probing, could you start using it for the compiler that already have it? (99% guarantee that we would forget to make this update when C++17 becomes the ROOT default).

Reply:
I already use it for compilers that have it, see lines above these :D